src/compiler/checker.ts

```diff
@@ -34086,6 +34197,438 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
+        return tryCreateTypeReference(getGlobalClassGetterDecoratorContextType(/*reportErrors*/ true), [thisType, valueType]);
+    }
+
+    function createClassSetterDecoratorContextType(thisType: Type, valueType: Type) {
+        return tryCreateTypeReference(getGlobalClassSetterDecoratorContextType(/*reportErrors*/ true), [thisType, valueType]);
+    }
+
+    function createClassAccessorDecoratorContextType(thisType: Type, valueType: Type) {
+        return tryCreateTypeReference(getGlobalClassAccessorDecoratorContextType(/*reportErrors*/ true), [thisType, valueType]);
+    }
+
+    function createClassFieldDecoratorContextType(thisType: Type, valueType: Type) {
+        return tryCreateTypeReference(getGlobalClassFieldDecoratorContextType(/*reportErrors*/ true), [thisType, valueType]);
+    }
+
+    function tryCreateClassMemberDecoratorContextOverrideType(nameType?: Type, isStatic?: boolean, isPrivate?: boolean) {
+        let members: SymbolTable | undefined;
+        if (nameType !== undefined) {
+            const symbol = createProperty("name" as __String, nameType);
+            members ??= createSymbolTable();
+            members.set(symbol.escapedName, symbol);
+        }
+        if (isPrivate !== undefined) {
+            const symbol = createProperty("private" as __String, isPrivate ? trueType : falseType);
+            members ??= createSymbolTable();
+            members.set(symbol.escapedName, symbol);
+        }
+        if (isStatic !== undefined) {
+            const symbol = createProperty("static" as __String, isStatic ? trueType : falseType);
+            members ??= createSymbolTable();
+            members.set(symbol.escapedName, symbol);
+        }
+        if (members) {
+            return createAnonymousType(/*symbol*/ undefined, members, emptyArray, emptyArray, emptyArray);
+        }
+    }
+
+    function createClassMemberDecoratorContextTypeForNode(node: MethodDeclaration | AccessorDeclaration | PropertyDeclaration, thisType: Type, valueType: Type) {
+        const isStatic = hasStaticModifier(node);
+        const isPrivate = isPrivateIdentifier(node.name);
+        const nameType = isPrivate ? getStringLiteralType(idText(node.name)) : getLiteralTypeFromPropertyName(node.name);
+        const contextType =
+            isMethodDeclaration(node) ? createClassMethodDecoratorContextType(thisType, valueType) :
+            isGetAccessorDeclaration(node) ? createClassGetterDecoratorContextType(thisType, valueType) :
+            isSetAccessorDeclaration(node) ? createClassSetterDecoratorContextType(thisType, valueType) :
+            isAutoAccessorPropertyDeclaration(node) ? createClassAccessorDecoratorContextType(thisType, valueType) :
+            isPropertyDeclaration(node) ? createClassFieldDecoratorContextType(thisType, valueType) :
+            Debug.failBadSyntaxKind(node);
+        const overrideType = tryCreateClassMemberDecoratorContextOverrideType(nameType, isStatic, isPrivate);

```

Comment:
This creates a new type identity every time it is called with no caching. That's going to create a lot of work downstream. Modulo overload resolution, when obtaining the contextual type for a particular node multiple times, we should always return the same type identity.

Reply:
I've added a cache for this, if you'd like to take another look.